File: src/csharp/core/src/Utils/Initials.cs
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace DiceBear.Internal
{
    /// <summary>
    /// Derives display initials from a seed string.
    /// </summary>
    /// <remarks>
    /// Words are split by Unicode letter and mark categories, the equivalent of
    /// the <c>\p{L}</c> / <c>\p{M}</c> classes the JS reference matches with,
    /// so the result matches the other ports for accented and non-Latin input.
    /// The scan walks code points rather than UTF-16 units because .NET's
    /// regular expressions classify a surrogate pair as two <c>Cs</c>
    /// characters, which would drop letters outside the Basic Multilingual
    /// Plane that the reference matches.
    /// See https://www.regular-expressions.info/unicode.html
    /// </remarks>
    internal static class Initials
    {
        /// <summary>
        /// Apostrophes and accents that should not break a word (e.g. l'eau,
        /// d´or).
        /// </summary>
        private const string Quotes = "`´'ʼ";

        /// <summary>
        /// Returns one or two uppercase initials for the given seed,
        /// discarding an "@…" suffix so email addresses yield a single
        /// initial.
        /// </summary>
        internal static string FromSeed(string seed) => FromSeed(seed, true);

        /// <summary>
        /// <see cref="FromSeed(string)"/> with control over whether the "@…"
        /// suffix is stripped before the initials are derived.
        /// </summary>
        private static string FromSeed(string seed, bool discardAtSymbol)
        {
            var input = seed;

            if (discardAtSymbol)
            {
                // Strip the whole @ suffix, including any line terminators.
                // Only the first @ matters since the strip reaches the end.
                var at = input.IndexOf('@');

                if (at >= 0)
                {
                    input = input.Substring(0, at);
                }
            }

            input = StripQuotes(input);

            var words = Words(input);

            if (words.Count == 0)
            {
                return discardAtSymbol ? FromSeed(seed, false) : string.Empty;
            }

            if (words.Count == 1)
            {
                var leading = LeadingUnits(words[0], 2);

                return leading.Length == 0 ? string.Empty : Uppercase.JsToUpperCase(leading);
            }

            var first = LeadingUnits(words[0], 1);
            var last = LeadingUnits(words[words.Count - 1], 1);

            if (first.Length == 0 || last.Length == 0)
            {
                return string.Empty;
            }

            return Uppercase.JsToUpperCase(first + last);
        }

        /// <summary>
        /// Removes the apostrophes and accents that must not break a word.
        /// </summary>
        private static string StripQuotes(string value)
        {
            var builder = new StringBuilder(value.Length);

            foreach (var ch in value)
            {
                if (Quotes.IndexOf(ch) < 0)
                {
                    builder.Append(ch);
                }
            }

            return builder.ToString();
        }

        /// <summary>
        /// Splits <paramref name="value"/> into words: a letter followed by any
        /// run of letters and combining marks.
        /// </summary>
        private static List<string> Words(string value)
        {
            var words = new List<string>();
            var index = 0;

            while (index < value.Length)
            {
                var width = Width(value, index);

                if (!IsLetter(value, index))
                {
                    index += width;
                    continue;
                }

                var start = index;
                index += width;

                while (index < value.Length && (IsLetter(value, index) || IsMark(value, index)))
                {
                    index += Width(value, index);
                }

                words.Add(value.Substring(start, index - start));
            }

            return words;
        }

        /// <summary>
        /// Returns the first <paramref name="max"/> letter-plus-marks units of
        /// a word, or fewer when the word is shorter.
        /// </summary>
        private static string LeadingUnits(string word, int max)
        {
            var index = 0;

            for (var taken = 0; taken < max && index < word.Length; taken++)
            {
                if (!IsLetter(word, index))
                {
                    break;
                }

                index += Width(word, index);

                while (index < word.Length && IsMark(word, index))
                {
                    index += Width(word, index);
                }
            }

            return word.Substring(0, index);
        }

        private static int Width(string value, int index) =>
            char.IsSurrogatePair(value, index) ? 2 : 1;

        private static bool IsLetter(string value, int index)
        {
            switch (CharUnicodeInfo.GetUnicodeCategory(value, index))
            {
                case UnicodeCategory.UppercaseLetter:
                case UnicodeCategory.LowercaseLetter:
                case UnicodeCategory.TitlecaseLetter:
                case UnicodeCategory.ModifierLetter:
                case UnicodeCategory.OtherLetter:
                    return true;
                default:
                    return false;
            }
        }

        private static bool IsMark(string value, int index)
        {
            switch (CharUnicodeInfo.GetUnicodeCategory(value, index))
            {
                case UnicodeCategory.NonSpacingMark:
                case UnicodeCategory.SpacingCombiningMark:
                case UnicodeCategory.EnclosingMark:
                    return true;
                default:
                    return false;
            }
        }
    }
}

File: src/rust/core/src/utils/initials.rs
//! Derives display initials from a seed string.
//!
//! Words are split by Unicode letter/mark classes (`\p{L}`, `\p{M}`) so the
//! result matches every other port for accented and non-Latin input.
//! See <https://www.regular-expressions.info/unicode.html>.

use std::sync::LazyLock;

use regex::Regex;

/// `@…` — the entire suffix (e.g. an email domain), stripped so the local part
/// yields one word. `(?s)` makes `.` match line terminators too, so the whole
/// tail is removed — matching the dotall `@.*` strip in every port.
static RE_AT: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?s)@.*").unwrap());
/// Apostrophes and accents that should not break a word (e.g. `l'eau`, `d´or`).
static RE_QUOTES: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[`´'ʼ]").unwrap());
/// A word: a letter followed by any letters or combining marks.
static RE_WORD: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\p{L}[\p{L}\p{M}]*").unwrap());
/// The first one or two grapheme-ish units (letter + trailing marks) of a word.
static RE_TWO: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^(?:\p{L}\p{M}*){1,2}").unwrap());
/// The first single unit (letter + trailing marks) of a word.
static RE_ONE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^(?:\p{L}\p{M}*)").unwrap());

/// Returns one or two uppercase initials for `seed`, discarding an `@…` suffix
/// so email addresses yield a single initial.
pub fn from_seed(seed: &str) -> String {
    from_seed_with(seed, true)
}

/// Like [`from_seed`], but `discard_at_symbol` controls whether an `@…` suffix
/// is stripped before deriving initials.
pub fn from_seed_with(seed: &str, discard_at_symbol: bool) -> String {
    let mut input = seed.to_string();

    if discard_at_symbol {
        // Only the first match, mirroring JS `replace(/@.*/, '')` (no `g` flag).
        input = RE_AT.replacen(&input, 1, "").into_owned();
    }

    input = RE_QUOTES.replace_all(&input, "").into_owned();

    let words: Vec<&str> = RE_WORD.find_iter(&input).map(|m| m.as_str()).collect();

    if words.is_empty() {
        return if discard_at_symbol {
            from_seed_with(seed, false)
        } else {
            String::new()
        };
    }

    if words.len() == 1 {
        return RE_TWO
            .find(words[0])
            .map(|m| m.as_str().to_uppercase())
            .unwrap_or_default();
    }

    let first = RE_ONE.find(words[0]);
    let last = RE_ONE.find(words[words.len() - 1]);

    match (first, last) {
        (Some(f), Some(l)) => format!("{}{}", f.as_str(), l.as_str()).to_uppercase(),
        _ => String::new(),
    }
}

/// Cross-language parity against the shared `initials.json` fixture.
#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::PathBuf;

    use serde_json::Value;

    use super::from_seed;

    fn fixture(name: &str) -> Value {
        let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("../../../tests/fixtures/parity")
            .join(name);
        let text =
            fs::read_to_string(&path).unwrap_or_else(|e| panic!("read {}: {e}", path.display()));
        serde_json::from_str(&text).unwrap()
    }

    #[test]
    fn initials_parity() {
        for e in fixture("initials.json").as_array().unwrap() {
            let seed = e["seed"].as_str().unwrap();
            assert_eq!(
                from_seed(seed),
                e["result"].as_str().unwrap(),
                "initials {seed:?}"
            );
        }
    }
}
